django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -0,0 +1,467 @@
+        package_qs = self._filter_by_query(params.get("q"), package_qs)
+        package_qs = self._order_queryset(params["ordering"], package_qs)
+        package_page = self._paginate(params, package_qs)
+        serializer = self._serialize_results(params, package_page)
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def get_queryset(self) -> QuerySet[Package]:
+        """
+        Implement QuerySet operations shared by all endpoints.
+        """
+
+        this_package = Package.objects.filter(pk=OuterRef("pk"))
+
+        return (
+            self._get_base_queryset()
+            .active()  # type: ignore
+            .prefetch_related(
+                "community_listings__categories",
+                "community_listings__community",
+            )
+            .select_related("latest", "namespace")
+            .annotate(
+                download_count=Subquery(
+                    this_package.annotate(
+                        downloads=Sum("versions__downloads"),
+                    ).values("downloads"),
+                ),
+            )
+            .annotate(
+                rating_count=Subquery(
+                    this_package.annotate(
+                        ratings=Count("package_ratings"),
+                    ).values("ratings"),
+                ),
+            )
+        )
+
+    def _filter_deprecated(
+        self,
+        show_deprecated: bool,
+        queryset: QuerySet[Package],
+    ) -> QuerySet[Package]:
+        """
+        Deprecated packages are included only if specifically requested.
+        """
+        if show_deprecated:
+            return queryset
+
+        return queryset.exclude(is_deprecated=True)
+
+    def _filter_nsfw(
+        self,
+        show_nsfw: bool,
+        queryset: QuerySet[Package],
+    ) -> QuerySet[Package]:
+        """
+        NSFW packages are included only if specifically requested.
+        """
+        if show_nsfw:
+            return queryset
+
+        return queryset.exclude(community_listings__has_nsfw_content=True)
+
+    def _filter_by_categories(
+        self,
+        params: OrderedDict,
+        queryset: QuerySet[Package],
+    ) -> QuerySet[Package]:
+        """
+        Include only packages (not) belonging to specific categories.
+
+        Multiple categories are OR-joined, i.e. if included_categories
+        contain A and B, packages belonging to either will be returned.
+        """
+        if params["included_categories"]:
+            queryset = queryset.exclude(
+                ~Q(
+                    community_listings__categories__slug__in=params[
+                        "included_categories"
+                    ],
+                ),
+            )
+
+        if params["excluded_categories"]:
+            queryset = queryset.exclude(
+                community_listings__categories__slug__in=params["excluded_categories"],
+            )
+
+        return queryset
```

Comment:
This probably works due to the implicit community scoping but it's wrong to assume category slugs are unique. Would prefer to see this use actual unique (and preferably immutable) IDs